dart/common/Uri.cpp

```diff
@@ -196,12 +206,125 @@ void Uri::clear()
 }
 
 //==============================================================================
+bool Uri::fromString(const std::string& _input)
+{
+  // This is regex is from Appendix B of RFC 3986.
+  static regex uriRegex(
+    R"END(^(([^:/?#]+):)?(//([^/?#]*))?([^?#]*)(\?([^#]*))?(#(.*))?)END",
+    regex::extended | regex::optimize);
+  static const size_t schemeIndex = 2;
+  static const size_t authorityIndex = 4;
+  static const size_t pathIndex = 5;
+  static const size_t queryIndex = 7;
+  static const size_t fragmentIndex = 9;
+
+  clear();
+
+  smatch matches;
+  if(!regex_match(_input, matches, uriRegex))
+    return false;
+
+  assert(matches.size() > schemeIndex);
+  const ssub_match& schemeMatch = matches[schemeIndex];
+  if(schemeMatch.matched)
+    mScheme = schemeMatch;
+
+  assert(matches.size() > authorityIndex);
+  const ssub_match& authorityMatch = matches[authorityIndex];
+  if(authorityMatch.matched)
+    mAuthority = authorityMatch;
+
+  assert(matches.size() > pathIndex);
+  const ssub_match& pathMatch = matches[pathIndex];
+  if(pathMatch.matched)
+  {
+    mPath = pathMatch;
+
+#ifdef _WIN32
+    if(mScheme && *mScheme == "file")
+    {
+      if(!mPath->empty() && mPath->at(0) == '/')
+        mPath = mPath->substr(1, mPath->size() - 1);
+    }
+#endif
+  }
```

Comment:
I was hoping to keep `Uri` as close to a mutable struct as possible. Many `ResourceRetreiver`s rewrite URIs to other types of URIs (e.g. resolve `package://` URIs to `file://` URIs), which is easier to implement if you have direct access to the URI components.

For example, a simple assignment like `*old_uri.schema = "file"` would turn into a verbose (and error-prone) constructor call that looks like this:

```c++
Uri new_uri("file", old_uri.getAuthority(), old_uri.getPath(), old_uri.getQuery(), old_uri.getFragment()
```

We also can't replace `UriComponent` with an `std::string`. The RFC explicitly states that it's important to differentiate between "missing" components and "empty" components, because they may serialize to very different meanings.